openai/src/commonMain/kotlin/com/xebia/functional/xef/conversation/llm/openai/OpenAI.kt

```diff
@@ -51,51 +57,60 @@ class OpenAI(internal var token: String? = null, internal var host: String? = nu
     }
   }
 
-  val GPT_4 by lazy { autoClose(OpenAIModel(this, "gpt-4", ModelType.GPT_4)) }
+  val defaultClient =
+    OpenAIClient(
+        host = getHost()?.let { OpenAIHost(it) } ?: OpenAIHost.OpenAI,
+        token = getToken(),
+        logging = LoggingConfig(LogLevel.None),
+        headers = mapOf("Authorization" to " Bearer ${getToken()}"),
+      )
+      .let { autoClose(it) }
 
-  val GPT_4_0314 by lazy { autoClose(OpenAIModel(this, "gpt-4-0314", ModelType.GPT_4)) }
+  val GPT_4 by lazy { autoClose(OpenAIChat(ModelType.GPT_4, defaultClient)) }
 
-  val GPT_4_32K by lazy { autoClose(OpenAIModel(this, "gpt-4-32k", ModelType.GPT_4_32K)) }
-
-  val GPT_3_5_TURBO by lazy {
-    autoClose(OpenAIModel(this, "gpt-3.5-turbo", ModelType.GPT_3_5_TURBO))
+  val GPT_4_0314 by lazy {
+    autoClose(OpenAIFunChat(ModelType.GPT_4, defaultClient)) // legacy
   }
 
+  val GPT_4_32K by lazy { autoClose(OpenAIChat(ModelType.GPT_4_32K, defaultClient)) }
+
+  val GPT_3_5_TURBO by lazy { autoClose(OpenAIChat(ModelType.GPT_3_5_TURBO, defaultClient)) }
+
   val GPT_3_5_TURBO_16K by lazy {
-    autoClose(OpenAIModel(this, "gpt-3.5-turbo-16k", ModelType.GPT_3_5_TURBO_16_K))
+    autoClose(OpenAIChat(ModelType.GPT_3_5_TURBO_16_K, defaultClient))
   }
 
   val GPT_3_5_TURBO_FUNCTIONS by lazy {
```

Comment:
It does not need to be addressed in this PR, but we saw there is a version gpt-4 that also accepts functions.

Reply:
YESSIRRRRR, as we say in German though _this is not my shoe_